### PR TITLE
experimental NTL support for cmake

### PR DESCRIPTION
**Pull request by dimpase**
for more knowledgeable in cmake people to look at. It does build on FreeBSD.

### CODE REVIEW

**Comment by wbhart** [comment]
@peter-frentrup

File: CMake/FindNTL.cmake [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  PATHS
+  ~/Library/Frameworks
+  /Library/Frameworks
+  /usr/local
+  /usr
+  /sw # Fink
+  /opt/local # DarwinPorts
+  /opt/csw # Blastwave
+  /opt
+)
```
Comment (isuruf) [inline CMake/FindNTL.cmake]: I would change this to,
```
FIND_PATH(NTL_INCLUDE_DIR NTL/RR.h
  HINTS
  $ENV{NTLDIR}
)
```

File: CMake/FindNTL.cmake [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  PATHS
+  ~/Library/Frameworks
+  /Library/Frameworks
+  /usr/local
+  /usr
+  /sw
+  /opt/local
+  /opt/csw
+  /opt
+)
```
Comment (isuruf) [inline CMake/FindNTL.cmake]:
```
FIND_LIBRARY(NTL_LIBRARY
  NAMES ntl
  HINTS
  $ENV{NTLDIR}
)
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -6,12 +6,13 @@ include(CheckFunctionExists)
...
 include(TestBigEndian)
 
-project(flint C)
+project(flint LANGUAGES C CXX)
 
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/CMake")
 
 find_package(GMP REQUIRED)
 find_package(MPFR REQUIRED)
+find_package(NTL REQUIRED)
```
Comment (isuruf) [inline CMakeLists.txt]: I'd make this optional as NTL with MPIR on windows 64-bit assumes LP64 and doesn't work.
Something like,
```
option(WITH_NTL "Build with NTL interface" OFF)
if (WITH_NTL)
    find_package(NTL REQUIRED)
endif()
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -182,16 +184,17 @@ endforeach()
 file(GLOB TEMP "${CMAKE_SOURCE_DIR}/*.h")
 list(APPEND HEADERS ${TEMP})
 
-add_library(flint ${SOURCES})
+add_library(flint SHARED ${SOURCES})
```
Comment (isuruf) [inline CMakeLists.txt]: This makes it impossible to build static libraries if needed. There's an option `cmake -DBUILD_SHARED_LIBS=on` which would build a shared library. (You could change the default value of off in BUILD_SHARED_LIBS to on if you like)

**Comment by dimpase** [comment]
Thanks - today I touched cmake configs for the 1st time - no wonder this is very far from prefect.

File: CMakeLists.txt (hunk 3 above)
Comment (dimpase) [reply CMakeLists.txt]: How does one check this ? Is it `cmake -D???` for some value of `???`.

Comment (isuruf) [reply CMakeLists.txt]: Yes, `cmake -DWITH_NTL=yes` would enable it.

**Comment by dimpase** [comment]
By the way, how do you test flint's cmake build?

**Comment by wbhart** [comment]
We don't. We had CI for a while, but it wasn't passing anyway, so we switched it off.

**Comment by dimpase** [comment]
cmake build does not seem to be working on Linux (Ubuntu 18) --- as well as on FreeBSD, in fact - I overlooked this) for me,
as libntl is not getting linked in:
```
$ ldd lib/libflint.so 
	linux-vdso.so.1 (0x00007ffe56374000)
	libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f0e59edf000)
	libmpfr.so.6 => /usr/lib/x86_64-linux-gnu/libmpfr.so.6 (0x00007f0e59c5f000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f0e59ad5000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0e59948000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f0e5992e000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0e59744000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f0e5aad2000)
```
whereas on the system I install libflint using apt and see
```
$ ldd /usr/lib/libflint.so 
	linux-vdso.so.1 (0x00007ffdbb11b000)
	libmpfr.so.6 => /usr/lib/x86_64-linux-gnu/libmpfr.so.6 (0x00007fb464072000)
	libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007fb463df1000)
	libntl.so.35 => /usr/lib/libntl.so.35 (0x00007fb4639d7000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fb4639b6000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fb46382c000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fb46369f000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fb463683000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fb463499000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fb464851000)
	libgf2x.so.1 => /usr/lib/x86_64-linux-gnu/libgf2x.so.1 (0x00007fb46328b000)
```

So there is a linking problem. What is this? A bug in Cmake? (As NTL is a C++ library, it's less obvious for cmake what to do). Same with cmake versions 3.12.1 and 3.13.4.

**Comment by isuruf** [comment]
https://github.com/dimpase/flint2/pull/1

**Comment by dimpase** [comment]
> We don't. We had CI for a while, but it wasn't passing anyway, so we switched it off.

After the update from @isuruf it seems to be possible to build and test with NTL on linux (latest Ubunti 18, and Gentoo) and FreeBSD, however tests get stuck on `fmpz_poly_factor-test-t-factor` (running for 40 minutes of CPU time). Is it a known problem with this particular test?

**Comment by isuruf** [comment]
You are probably building in Debug mode. Did you try `cmake -DCMAKE_BUILD_TYPE=Release` ?

**Comment by wbhart** [comment]
It tries to factor some huge polynomials. I think @isuruf is right.

I must speed that code up some day.

**Comment by dimpase** [comment]
> You are probably building in Debug mode. Did you try `cmake -DCMAKE_BUILD_TYPE=Release` ?

OK, this works, thanks.

**Comment by dimpase** [comment]
I've added https://github.com/wbhart/flint2/pull/543/commits/11a4f5d5465ccc1061970339865a638d68712fe0 to allow building on Gentoo, where GMP headers have a different layout to what's programmed in `FindGMP.cmake`. This certainly is far from ideal, but seems to make sense.

**Comment by isuruf** [comment]
https://github.com/dimpase/flint2/pull/2 changes some default options and fixes some things

**Comment by dimpase** [comment]
> We don't. We had CI for a while, but it wasn't passing anyway, so we switched it off.

here is a passing with tests and NTL cmake Travis build: https://travis-ci.com/dimpase/flint2/builds/101585015 from https://github.com/dimpase/flint2/tree/travis_cmake 

Would you like this merged into this PR?

**Comment by wbhart** [comment]
Yes, go ahead and put it into this PR. Thanks for your efforts.

**Comment by isuruf** [comment]
@dimpase, do you mind adding a cmake job in addition to the other CI jobs?

**Comment by dimpase** [comment]
> @dimpase, do you mind adding a cmake job in addition to the other CI jobs?

yes, that's what I am going to do, add things in my experimental branch to the pull request (naturally, adding them to the other CI things)

**Comment by isuruf** [comment]
Thanks. I've got a branch that fixes cmake builds on windows CI (except for 4 tests) on top of this branch. I'll send that when this PR is merged.

**Comment by dimpase** [comment]
OK, I think I'm done with this, please review.

**Comment by dimpase** [comment]
Probably Travis script may be made more clear by using their "matrix" stuff, but this can be done later.

**Comment by wbhart** [comment]
Thanks. I'll wait for @isuruf to sign off on it, then I'll merge it. I'm virtually ignorant about CMake.

**Pull request merged**